File: scripts/xlsx_to_json.py
#!/usr/bin/env python3
import sys
import os

try:
    import pandas as pd
except ImportError:
    print("Erro: A biblioteca 'pandas' não está instalada.")
    print("Instale executando: pip install pandas openpyxl")
    sys.exit(1)

def convert_xlsx_to_json(xlsx_path, json_path, sheet_name=0):
    if not os.path.exists(xlsx_path):
        print(f"Erro: Arquivo Excel não encontrado: {xlsx_path}")
        sys.exit(1)
        
    try:
        # Lê o Excel usando pandas
        # openpyxl é necessário para ler arquivos .xlsx no pandas
        df = pd.read_excel(xlsx_path, sheet_name=sheet_name)
        
        # Converte para JSON
        # orient='records' exporta como uma lista de objetos: [{col1: val1, col2: val2}, ...]
        df.to_json(json_path, orient='records', force_ascii=False, indent=4)
        print(f"✓ Sucesso! '{xlsx_path}' [Aba: {sheet_name}] convertido para '{json_path}'")
    except Exception as e:
        print(f"Erro na conversão: {e}")
        print("Certifique-se de que a biblioteca 'openpyxl' está instalada: pip install openpyxl")
        sys.exit(1)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Uso:")
        print("  1. Listar abas da planilha:")
        print("     python3 scripts/xlsx_to_json.py <arquivo_excel.xlsx>")
        print("  2. Converter uma aba específica para JSON (nome automático da aba):")
        print("     python3 scripts/xlsx_to_json.py <arquivo_excel.xlsx> <nome_ou_indice_da_aba>")
        print("  3. Converter TODAS as abas para arquivos JSON individuais:")
        print("     python3 scripts/xlsx_to_json.py <arquivo_excel.xlsx> all")
        print("  4. Converter com caminho de destino personalizado:")
        print("     python3 scripts/xlsx_to_json.py <arquivo_excel.xlsx> <arquivo_destino.json> <nome_ou_indice_da_aba>")
        sys.exit(1)
        
    xlsx_file = sys.argv[1]
    
    if not os.path.exists(xlsx_file):
        print(f"Erro: Arquivo Excel não encontrado: {xlsx_file}")
        sys.exit(1)
        
    # Se forneceu apenas o arquivo Excel, lista as abas disponíveis
    if len(sys.argv) == 2:
        try:
            xl = pd.ExcelFile(xlsx_file)
            print(f"\nPlanilhas (abas) encontradas em '{xlsx_file}':")
            for idx, name in enumerate(xl.sheet_names):
                print(f"  [{idx}] {name}")
            print("\nPara converter uma aba específica, execute:")
            print(f"  python3 scripts/xlsx_to_json.py \"{xlsx_file}\" \"NomeDaAba\"")
            print("Para converter todas as abas de uma vez, execute:")
            print(f"  python3 scripts/xlsx_to_json.py \"{xlsx_file}\" all")
        except ImportError:
            print("Erro: A biblioteca 'pandas' ou 'openpyxl' não está instalada.")
            print("Instale executando: pip install pandas openpyxl")
        except Exception as e:
            print(f"Erro ao ler as abas: {e}")
        sys.exit(0)
        
    # Tratamento para 2 ou mais argumentos
    if len(sys.argv) == 3:
        # Se o segundo argumento é "all", converte todas as abas
        if sys.argv[2].lower() == 'all':
            try:
                xl = pd.ExcelFile(xlsx_file)
                sheet_names = xl.sheet_names
            except Exception as e:
                print(f"Erro ao ler as abas: {e}")
                sys.exit(1)
                
            print(f"\nConvertendo todas as abas de '{xlsx_file}'...")
            excel_dir = os.path.dirname(xlsx_file)
            for sheet_name in sheet_names:
                sanitized = "".join(c for c in sheet_name if c not in '\\/:*?"<>|')
                json_file = os.path.join(excel_dir, f"{sanitized}.json")
                convert_xlsx_to_json(xlsx_file, json_file, sheet_name)
            print("\n✓ Concluído! Todas as abas foram convertidas com sucesso.")
            sys.exit(0)
            
        # Se o segundo argumento termina com .json, é o arquivo de destino
        elif sys.argv[2].endswith('.json'):
            json_file = sys.argv[2]
            sheet = 0
        else:
            # Caso contrário, é o nome/índice da aba e geramos o JSON na mesma pasta
            target_sheet = sys.argv[2]
            
            try:
                xl = pd.ExcelFile(xlsx_file)
                sheet_names = xl.sheet_names
            except Exception as e:
                print(f"Erro ao ler as abas: {e}")
                sys.exit(1)
                
            resolved_sheet_name = target_sheet
            if target_sheet.isdigit():
                sheet_idx = int(target_sheet)
                if sheet_idx < len(sheet_names):
                    resolved_sheet_name = sheet_names[sheet_idx]
                else:
                    print(f"Erro: Aba de índice {sheet_idx} não encontrada.")
                    sys.exit(1)
            else:
                if target_sheet not in sheet_names:
                    print(f"Erro: Aba '{target_sheet}' não encontrada.")
                    sys.exit(1)
                    
            # Sanatizar nome da aba para nome do arquivo
            sanitized = "".join(c for c in resolved_sheet_name if c not in '\\/:*?"<>|')
            excel_dir = os.path.dirname(xlsx_file)
            json_file = os.path.join(excel_dir, f"{sanitized}.json")
            sheet = resolved_sheet_name
    else:
        json_file = sys.argv[2]
        sheet = sys.argv[3]
        
    # Se sheet for um número, converte para int
    if isinstance(sheet, str) and sheet.isdigit():
        sheet = int(sheet)
        
    convert_xlsx_to_json(xlsx_file, json_file, sheet)

File: scripts/xlsx_to_json.js
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';

// Para poder verificar os argumentos da linha de comando
const args = process.argv.slice(2);

if (args.length < 1) {
  console.log("Uso:");
  console.log("  1. Listar abas da planilha:");
  console.log("     node scripts/xlsx_to_json.js <arquivo_excel.xlsx>");
  console.log("  2. Converter uma aba específica para JSON (nome automático da aba):");
  console.log("     node scripts/xlsx_to_json.js <arquivo_excel.xlsx> <nome_ou_indice_da_aba>");
  console.log("  3. Converter TODAS as abas para arquivos JSON individuais:");
  console.log("     node scripts/xlsx_to_json.js <arquivo_excel.xlsx> all");
  console.log("  4. Converter com caminho de destino personalizado:");
  console.log("     node scripts/xlsx_to_json.js <arquivo_excel.xlsx> <arquivo_destino.json> <nome_ou_indice_da_aba>");
  process.exit(1);
}

const xlsxPath = args[0];

// Importação dinâmica do 'xlsx' para poder exibir uma mensagem amigável caso não esteja instalado
try {
  const { default: XLSX } = await import('xlsx');
  
  if (!fs.existsSync(xlsxPath)) {
    console.error(`Erro: Arquivo Excel não encontrado: ${xlsxPath}`);
    process.exit(1);
  }

  // Lê o arquivo Excel
  const workbook = XLSX.readFile(xlsxPath);

  // Se forneceu apenas o arquivo Excel, lista as abas disponíveis
  if (args.length === 1) {
    console.log(`\nPlanilhas (abas) encontradas em '${xlsxPath}':`);
    workbook.SheetNames.forEach((name, idx) => {
      console.log(`  [${idx}] ${name}`);
    });
    console.log("\nPara converter uma aba específica, execute:");
    console.log(`  node scripts/xlsx_to_json.js "${xlsxPath}" "NomeDaAba"`);
    console.log("Para converter todas as abas de uma vez, execute:");
    console.log(`  node scripts/xlsx_to_json.js "${xlsxPath}" all`);
    process.exit(0);
  }

  const excelDir = path.dirname(xlsxPath);

  // Se o usuário quer converter TODAS as abas
  if (args.length === 2 && args[1].toLowerCase() === 'all') {
    console.log(`\nConvertendo todas as abas de '${xlsxPath}'...`);
    workbook.SheetNames.forEach(sheetName => {
      const worksheet = workbook.Sheets[sheetName];
      const jsonData = XLSX.utils.sheet_to_json(worksheet);
      
      const sanitizedSheetName = sheetName.replace(/[\\/:*?"<>|]/g, "_");
      const jsonPath = path.join(excelDir, `${sanitizedSheetName}.json`);
      
      fs.writeFileSync(jsonPath, JSON.stringify(jsonData, null, 2), 'utf-8');
      console.log(`  ✓ Aba [${sheetName}] convertida para '${jsonPath}'`);
    });
    console.log("\n✓ Concluído! Todas as abas foram convertidas com sucesso.");
    process.exit(0);
  }

  let jsonPath;
  let targetSheet;

  if (args.length === 2) {
    // Se o segundo argumento terminar com .json, trata como arquivo de destino personalizado e assume aba 0
    if (args[1].endsWith('.json')) {
      jsonPath = args[1];
      targetSheet = 0;
    } else {
      // Caso contrário, trata como o nome ou índice da aba, e o destino será automático
      targetSheet = args[1];
      
      // Resolve o nome real da aba para usá-lo como nome do arquivo JSON
      let resolvedSheetName = targetSheet;
      if (!isNaN(targetSheet)) {
        const sheetIndex = parseInt(targetSheet, 10);
        resolvedSheetName = workbook.SheetNames[sheetIndex];
        if (!resolvedSheetName) {
          console.error(`Erro: Aba de índice ${sheetIndex} não encontrada. Abas disponíveis: ${workbook.SheetNames.join(', ')}`);
          process.exit(1);
        }
      } else {
        if (!workbook.SheetNames.includes(targetSheet)) {
          console.error(`Erro: Aba '${targetSheet}' não encontrada. Abas disponíveis: ${workbook.SheetNames.join(', ')}`);
          process.exit(1);
        }
      }
      
      const sanitizedSheetName = resolvedSheetName.replace(/[\\/:*?"<>|]/g, "_");
      jsonPath = path.join(excelDir, `${sanitizedSheetName}.json`);
      targetSheet = resolvedSheetName;
    }
  } else {
    // Caso tenha 3 ou mais argumentos (ex: node script.js planilha.xlsx destino.json "Minha Aba")
    jsonPath = args[1];
    targetSheet = args[2];
  }

  // Resolve qual aba ler (por índice ou por nome) para a conversão
  let sheetName = targetSheet;
  if (!isNaN(targetSheet)) {
    const sheetIndex = parseInt(targetSheet, 10);
    sheetName = workbook.SheetNames[sheetIndex];
    if (!sheetName) {
      console.error(`Erro: Aba de índice ${sheetIndex} não encontrada. Abas disponíveis: ${workbook.SheetNames.join(', ')}`);
      process.exit(1);
    }
  } else {
    if (!workbook.SheetNames.includes(targetSheet)) {
      console.error(`Erro: Aba '${targetSheet}' não encontrada. Abas disponíveis: ${workbook.SheetNames.join(', ')}`);
      process.exit(1);
    }
  }

  const worksheet = workbook.Sheets[sheetName];
  
  // Converte a aba para JSON (lista de objetos)
  const jsonData = XLSX.utils.sheet_to_json(worksheet);
  
  // Salva no arquivo de destino
  fs.writeFileSync(jsonPath, JSON.stringify(jsonData, null, 2), 'utf-8');
  console.log(`✓ Sucesso! '${xlsxPath}' [Aba: ${sheetName}] convertido para '${jsonPath}'`);
} catch (err) {
  if (err.code === 'ERR_MODULE_NOT_FOUND' || err.message.includes('Cannot find package')) {
    console.error("Erro: A biblioteca 'xlsx' (SheetJS) não está instalada no projeto.");
    console.error("Instale-a executando:");
    console.error("  npm install xlsx");
  } else {
    console.error("Erro ao converter:", err.message);
  }
  process.exit(1);
}
